server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationDiagnosticsService.java

```diff
@@ -298,22 +305,173 @@ private CoordinationDiagnosticsResult diagnoseOnHaveNotSeenMasterRecently(Master
+                );
+            }
         }
         return result;
     }
 
+    /**
+     * This method checks whether each master eligible node has discovered each of the other master eligible nodes. If any node in
+     * nodeToClusterFormationStateMap to discover any other node in masterEligibleNodes, then this method returns true. If no discovery
+     * problems are detected, this method returns false.
+     * @param masterEligibleNodes The collection of all master eligible nodes
+     * @param nodeToClusterFormationStateMap A map of each master node to its ClusterFormationState
+     * @return true if there are discovery problems, false otherwise
+     */
+    private boolean anyNodeInClusterReportsDiscoveryProblems(
+        Collection<DiscoveryNode> masterEligibleNodes,
+        Map<DiscoveryNode, ClusterFormationFailureHelper.ClusterFormationState> nodeToClusterFormationStateMap
+    ) {
+        Map<DiscoveryNode, Collection<DiscoveryNode>> nodesNotDiscoveredMap = new HashMap<>();
+        for (Map.Entry<DiscoveryNode, ClusterFormationFailureHelper.ClusterFormationState> entry : nodeToClusterFormationStateMap
+            .entrySet()) {
+            Set<DiscoveryNode> foundPeersOnNode = new HashSet<>(entry.getValue().foundPeers());
+            if (foundPeersOnNode.containsAll(masterEligibleNodes) == false) {
+                Collection<DiscoveryNode> nodesNotDiscovered = masterEligibleNodes.stream()
+                    .filter(node -> foundPeersOnNode.contains(node) == false)
+                    .toList();
+                nodesNotDiscoveredMap.put(entry.getKey(), nodesNotDiscovered);
+            }
+        }
+        if (nodesNotDiscoveredMap.isEmpty()) {
+            return false;
+        } else {
+            String nodeDiscoveryProblemsMessage = nodesNotDiscoveredMap.entrySet()
+                .stream()
+                .map(
+                    entry -> String.format(
+                        Locale.ROOT,
+                        "%s cannot discover [%s]",
+                        entry.getKey().getName(),
+                        entry.getValue().stream().map(DiscoveryNode::getName).collect(Collectors.joining(", "))
+                    )
+                )
+                .collect(Collectors.joining("; "));
+            logger.info("The following nodes report discovery problems: {}", nodeDiscoveryProblemsMessage);
+            return true;
+        }
+    }
+
+    /**
+     * This method checks that each master eligible node in the quorum thinks that it can form a quorum. If there are nodes that report a
+     * problem forming a quorum, this method returns true
+     * @param nodeToClusterFormationStateMap A map of each master node to its ClusterFormationState
+     * @return True if any nodes in nodeToClusterFormationStateMap report a problem forming a quorum, false otherwise.
+     */
+    private boolean anyNodeInClusterReportsQuorumProblems(
```

Comment:
I think this is a bit confusing and would benefit from possibly being static (if possible), package protected and unit tested.

`quorumProblems` is very generic  - would it be that not enough nodes are present to create a quorum? Or they can't talk to the nodes in order to form a quorum?

Reply:
All we know is that there are quorum problems. We are not reporting on the reason at this time (that's probably one of the things we'll put in the details in the near future).
This one is currently unit tested by `testRedForNoMasterNoQuorum`. Do you think there's a need to crack into the private method?

Reply:
I think a failure in `testRedForNoMasterNoQuorum` could come from many sources as it goes through `diagnoseMaster...` (which is ok for integration tests) but for unit tests it'd be preferable  to have an immediate cause (and unit tests could serve for documentation as to what the method is looking for)

Reply:
This one has its own unit test. I've also added to the javadocs and unit test comments to try to make it more clear how it is looking for quorum problems.